### DISCUSSION TITLE
WasmEdge supports passing complex struct data type already?

### DISCUSSION THREAD

**Discussion by KeeganJin** [WasmEdge/WasmEdge#2830] [Q&A]
Hi all,
Last time I checked the [landing page](https://wasmedge.org/docs/category/passing-complex-data), I found WasmEdge supported Scalar types, String, Vec as Parameter and Scalar types, String, Vec, Tuple, and Result as return values.

But I surprisingly found that it already supported returning complex struct data types (nested) after checking out the [wasmedge-bindgen examples](https://github.com/second-state/wasmedge-rustsdk-examples/blob/main/wasmedge-bindgen/wasm-lib/src/lib.rs). Thanks for the great work. 

Is passing complex struct data types(nested) as parameters also supported? Thanks :>

**Comment by hydai** [comment]
Hi @apepkuss 
Can you help answer this?

**Comment by apepkuss** [comment]
@KeeganJin The answer is Yes. Suppose that you have a complex type named `Point`, and wanna pass an instance of this type to a host function.

```rust
struct Point {
    x: f32,
    y: f32,
}
```
What you need to do is to implement `serde::{Deserialize, Serialize}` traits:

```rust
#[derive(Serialize, Deserialize, Debug)]
struct Point {
    x: f32,
    y: f32,
}
```

(just as shown in this [code](https://github.com/second-state/wasmedge-rustsdk-examples/blob/main/wasmedge-bindgen/wasm-lib/src/lib.rs#L8C1-L12))

Each time you wanna pass a value of `Point` type to a target host function, just serialize the value to a string, and then pass the string to the function (just as shown in this [code](https://github.com/second-state/wasmedge-rustsdk-examples/blob/main/wasmedge-bindgen/call-wasm-lib/src/main.rs#L53). In the host function, you just deserialize the string to the original value of `Point` type first, then you can use it in the host function.

Hope the explanation above can help you! Thanks a lot!

**Reply by dmd17** [reply]
@apepkuss when will passing complex data types be supported in C/C++ SDK ?

**Reply by hydai** [reply]
Hi @dmd17 
Currently, we have yet to make a plan for supporting the complex types in C/C++ SDK. The previous bindgen is based on wasm-bindgen. However, the generated output is unstable and may break when bumping the version. Unless there is a contributor who would like to take over maintaining the synchronization, this is no longer on our milestones.

**Reply by dmd17** [reply]
Thank you. I'll have a look. The previous bindgen is which? this one https://github.com/second-state/wasm-bindgen ?

related to this question, how can we pass a binary blob from native in C to wasm?

**Reply by dannypsnl** [reply]
I think you can use `WasmEdge_MemoryInstanceGetPointer` or `WasmEdge_MemoryInstanceSetData`.

**Comment by apepkuss** [comment]
@hydai Could you please help answer the question about C/C++ SDK from @dmd17? Thanks a lot!
